src/GovUk.Education.ExploreEducationStatistics.Data.Services/DataService.cs

```diff
@@ -1,65 +1,64 @@
         {
             _resultBuilder = resultBuilder;
             _subjectMetaService = subjectMetaService;
-            _logger = logger;
         }
 
-        public override ResultWithMetaViewModel Query(ObservationQueryContext queryContext, ReleaseId? releaseId = null)
+        public override Task<Either<ActionResult, ResultWithMetaViewModel>> Query(ObservationQueryContext queryContext)
         {
-            var stopwatch = Stopwatch.StartNew();
-            stopwatch.Start();
+            return CheckSubjectExists(queryContext.SubjectId)
+                .OnSuccess(CheckCanViewSubjectData)
+                .OnSuccess(_ =>
+                {
+                    var observations = GetObservations(queryContext).AsQueryable();
 
-            var observations = GetObservations(queryContext, releaseId).AsQueryable();
+                    if (!observations.Any())
+                    {
+                        return Task.FromResult(
+                            new Either<ActionResult, ResultWithMetaViewModel>(new ResultWithMetaViewModel()));
+                    }
 
-            _logger.LogTrace("Got Observations in {Time} ms", stopwatch.Elapsed.TotalMilliseconds);
-            stopwatch.Restart();
+                    var result = observations
+                        .Select(observation => _resultBuilder.BuildResult(observation, queryContext.Indicators))
+                        .ToList();
 
-            if (!observations.Any())
-            {
-                return new ResultWithMetaViewModel();
-            }
-
-            var result = observations
-                .Select(observation => _resultBuilder.BuildResult(observation, queryContext.Indicators))
-                .ToList();
-
-            _logger.LogTrace("Built Observation results in {Time} ms", stopwatch.Elapsed.TotalMilliseconds);
-            stopwatch.Restart();
-
-            var metaData = _subjectMetaService.GetSubjectMeta(SubjectMetaQueryContext.FromObservationQueryContext(queryContext), observations);
-
-            _logger.LogTrace("Got meta data for Observations in {Time} ms", stopwatch.Elapsed.TotalMilliseconds);
-            stopwatch.Stop();
-
-            return new ResultWithMetaViewModel
-            {
-                Result = result,
-                MetaData = metaData
-            };
+                    return _subjectMetaService.GetSubjectMeta(SubjectMetaQueryContext.FromObservationQueryContext(queryContext), observations)
+                        .OnSuccess(subjectMetaViewModel => new ResultWithMetaViewModel
+                    {
+                        MetaData = subjectMetaViewModel,
+                        Result = result
+                    });
+                });
+        }
+        
+        private async Task<Either<ActionResult, bool>> CheckCanViewSubjectData(Subject subject)
+        {
+            var result = subject.Release.Live || await _userService.MatchesPolicy(subject, CanViewSubjectData);
```

Comment:
With other auth handler scenarios when the status of a release is important to the outcome of the permission check, I put the release status logic into the auth handler itself e.g. `UpdateSpecificReleaseHasUpdaterRoleOnReleaseAuthorizationHandler` (there's a mouthful!) checks that not only does a user have an "editable" role on the Release, but also that the release is not yet Approved.

Would it be good to follow the same pattern here i.e. bake that status check into the auth handler itself, rather than have that logic in 2 different places?

Reply:
That's done now, check out ViewSubjectDataForSpecificReleaseAuthorizationHandler